### PR TITLE
Added support for deleting rights on resources from resourceregistry

### PR DESCRIPTION
**Pull request by Sondre-B**
…in the DeleteRules operation.

## Description
Added support for deleting rights on resources from resourceregistry in the DeleteRulesOperation. It is now posible to pass ResourceRegistryId for a resource to have the rights revoked.

## Related Issue(s)
- #150

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 1]
```diff
@@ -405,14 +423,11 @@ private async Task<List<Rule>> ProcessPolicyFile(string policyPath, string org,
 
         private async Task<List<Rule>> DeleteAllRulesInPolicy(RequestToDelete policyToDelete)
         {
+            string altinnAppId = null;
+            ServiceResource resource = null;
             DelegationHelper.TryGetResourceFromAttributeMatch(policyToDelete.PolicyMatch.Resource, out ResourceAttributeMatchType resourceMatchType, out string resourceRegistryId, out string org, out string app);
             string coveredBy = DelegationHelper.GetCoveredByFromMatch(policyToDelete.PolicyMatch.CoveredBy, out int? coveredByUserId, out int? coveredByPartyId);
 
-            if (resourceMatchType != ResourceAttributeMatchType.AltinnApp)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:411]: Still should check the result of the TryGetResourceFromAttributeMatch operation. If it fails and/or the resourceMatchType is none operation should fail as the resource cannot be identified

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 2]
```diff
@@ -518,11 +546,6 @@ private async Task<List<Rule>> DeleteRulesInPolicy(RequestToDelete rulesToDelete
 
             DelegationHelper.TryGetResourceFromAttributeMatch(rulesToDelete.PolicyMatch.Resource, out ResourceAttributeMatchType resourceMatchType, out string resourceRegistryId, out string org, out string app);
 
-            if (resourceMatchType != ResourceAttributeMatchType.AltinnApp)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:521]: Same here as in DeleteAllRulesInPolicy:
Still should check the result of the TryGetResourceFromAttributeMatch operation. If it fails and/or the resourceMatchType is none operation should fail as the resource cannot be identified

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 3]
```diff
@@ -405,14 +423,11 @@ private async Task<List<Rule>> ProcessPolicyFile(string policyPath, string org,
...
             DelegationHelper.TryGetResourceFromAttributeMatch(policyToDelete.PolicyMatch.Resource, out ResourceAttributeMatchType resourceMatchType, out string resourceRegistryId, out string org, out string app);
             string coveredBy = DelegationHelper.GetCoveredByFromMatch(policyToDelete.PolicyMatch.CoveredBy, out int? coveredByUserId, out int? coveredByPartyId);
 
-            if (resourceMatchType != ResourceAttributeMatchType.AltinnApp)
-            {
-                throw new NotImplementedException("Deletion of rules for this resource type is not implemented");
-            }
-
             string policyPath;
             try
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:432]: LogError for catched exception for this try is still org/app specific

**Review by jonkjetiloye — CHANGES_REQUESTED** [review]

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Data/blobs/input/resourceregistry/resource2/50001337/u20001337/delegationpolicy.xml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xacml:Policy PolicyId="urn:altinn:policyid:1" Version="1.0" RuleCombiningAlgId="urn:oasis:names:tc:xacml:3.0:policy-combining-algorithm:deny-overrides" xmlns:xacml="urn:oasis:names:tc:xacml:3.0:core:schema:wd-17">
+	<xacml:Description>Delegation policy containing all delegated rights/actions from 50001337 to 20001337, for resources on the app; org1/app3</xacml:Description>
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Data/blobs/input/resourceregistry/resource2/50001337/u20001337/delegationpolicy.xml:3]: copy/pasted description

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Altinn.AuthorizationAdmin.Tests.csproj [hunk 5]
```diff
@@ -66,5 +69,11 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </None>
   </ItemGroup>
+
+  <ItemGroup>
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Altinn.AuthorizationAdmin.Tests.csproj:73]: Please remove all project file clutter for individual files under the Data folder, which should be unnecessary and already covered by the catch all item group.

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs [hunk 6]
```diff
@@ -115,6 +115,49 @@ public async Task Post_DeleteRules_Success()
...
+        public async Task Post_DeleteResourceRegistryRules_Success()
+        {
+            // Arrange
+            Stream dataStream = File.OpenRead("Data/Json/DeleteRules/ReadResource2_50001337_20001337.json");
+            StreamContent content = new StreamContent(dataStream);
+            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
+
+            List<Rule> expected = new List<Rule>
+            {
+                TestDataUtil.GetRuleModel(20001336, 50001337, "20001337", AltinnXacmlConstants.MatchAttributeIdentifiers.UserAttribute, "read", null, null, createdSuccessfully: true, resourceRegistryId: "resource2"),
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs:141]: createdSuccessfully is irrelevant for a delete operation. 
Should set the expected ruleIds which are to be deleted for verification against the response

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs [hunk 7]
```diff
@@ -115,6 +115,49 @@ public async Task Post_DeleteRules_Success()
...
+        /// Calling the POST operation for DeleteRules to perform a valid deletion
+        /// Input:
+        /// List of two one rule in one policy for deletion of the resource between for a single offeredby/coveredby combination resulting in a single policyfile beeing updated.
+        /// Expected Result:
+        /// Rules are deleted and returned with the CreatedSuccessfully flag set and rule ids
+        /// Success Criteria:
+        /// DeleteRules returns status code 201 and list of rules deleted to match expected
+        /// </summary>
+        [Fact]
+        public async Task Post_DeleteResourceRegistryRules_Success()
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs:132]: should also add test for DeletePolicy

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 8]
```diff
@@ -309,12 +311,23 @@ private async Task<List<Rule>> ProcessPolicyFile(string policyPath, string org,
                 return null;
             }
 
+            if (!string.IsNullOrWhiteSpace(org) || !string.IsNullOrWhiteSpace(app))
+            {
+                altinnAppId = $"{org}/{app}";
+            }
+
+            resource = await _resourceRegistryClient.GetResource(resourceregistryId);
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:319]: Resource lookup should only be performed if resourceRegistryId has a value

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 9]
```diff
@@ -437,9 +466,21 @@ private async Task<List<Rule>> DeleteAllRulesInPolicy(RequestToDelete policyToDe
                 return null;
             }
 
+            if (resourceMatchType == ResourceAttributeMatchType.AltinnApp)
+            {
+                altinnAppId = $"{org}/{app}";
+            }
+
+            resource = await _resourceRegistryClient.GetResource(resourceRegistryId);
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:474]: same here, use resourceMatchType.ResourceRegistry for specific logic as _resourceRegistryClient should not be used for altinn app scenario

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs (hunk 7 above)
Comment (jonkjetiloye) [reply backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs:132]: Should also have negative tests? E.g.
- Delete request without valid resource must fail

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs (hunk 8 above)
Comment (jonkjetiloye) [reply backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:319]: but also unsure whether we need to actually lookup the resource in the registry for delete operation. 
Should perhaps be enough that the delegation policy exist and we are able to get a lease lock for the blob?

**Comment by Sondre-B** [comment]
Closing this PR, replaced by https://github.com/Altinn/altinn-access-management/pull/170.

**Pull request closed without merging**